File: expectanim/src/main/java/com/github/florent37/expectanim/core/position/PositionAnimExpectationSameCenterAs.java
package com.github.florent37.expectanim.core.position;

import android.view.View;

/**
 * Created by florentchampigny on 17/02/2017.
 */

public class PositionAnimExpectationSameCenterAs extends PositionAnimationViewDependant {

    private final boolean horizontal;
    private final boolean vertical;

    public PositionAnimExpectationSameCenterAs(View otherView, boolean horizontal, boolean vertical) {
        super(otherView);
        setForPositionX(true);
        setForPositionY(true);
        this.horizontal = horizontal;
        this.vertical = vertical;
    }

    @Override
    public Float getCalculatedValueX(View viewToMove) {
        if(horizontal) {

            final float x = viewCalculator.finalPositionLeftOfView(otherView);
            final float myWidth = viewToMove.getWidth() / 2f;
            final float hisWidth = viewCalculator.finalWidthOfView(otherView) / 2f;

            if (myWidth > hisWidth) {
                return x - myWidth + hisWidth;
            } else {
                return x - hisWidth + myWidth;
            }
        } else return null;
    }

    @Override
    public Float getCalculatedValueY(View viewToMove) {
        if(vertical) {

            final float y = viewCalculator.finalPositionTopOfView(otherView);
            final float myHeight = viewToMove.getHeight() / 2f;
            final float hisHeight = viewCalculator.finalHeightOfView(otherView) / 2f;

            if (myHeight > hisHeight) {
                return y + myHeight - hisHeight;
            } else {
                return y + hisHeight - myHeight;
            }

        } else return null;
    }

}

File: kotlinpleaseanimate/src/main/java/com/github/florent37/kotlin/pleaseanimate/core/position/PositionAnimExpectationSameCenterAs.kt
package com.github.florent37.kotlin.pleaseanimate.core.position

import android.view.View

class PositionAnimExpectationSameCenterAs(otherView: View, private val horizontal: Boolean, private val vertical: Boolean) : PositionAnimationViewDependant(otherView) {

    init {
        isForPositionX = true
        isForPositionY = true
    }

    override fun getCalculatedValueX(viewToMove: View): Float? {
        if (horizontal) {
            val x = viewCalculator!!.finalPositionLeftOfView(otherView)
            val myWidth = viewToMove.width / 2f
            val hisWidth = viewCalculator!!.finalWidthOfView(otherView) / 2f

            return if (myWidth > hisWidth) {
                x - myWidth + hisWidth
            } else {
                x - hisWidth + myWidth
            }
        } else
            return null
    }

    override fun getCalculatedValueY(viewToMove: View): Float? {
        if (vertical) {
            val y = viewCalculator!!.finalPositionTopOfView(otherView)
            val myHeight = viewToMove.height / 2f
            val hisHeight = viewCalculator!!.finalHeightOfView(otherView) / 2f

            return if (myHeight > hisHeight) {
                y + myHeight - hisHeight
            } else {
                y + hisHeight - myHeight
            }
        } else
            return null
    }

}
